### PR TITLE
Set transaction protocol for optimizing

### PR DESCRIPTION
**Pull request by osopardo1**
This PR is solving #41 and adding the logic on Qbeast-Spark to interact with the Keeper. 

There's two main changes on Indexed Table:

- In the **writing** part, it would try to commit and check if some conflict error has been thrown. In the affirmative case, it would need to understand if the conflict is unsolvable or not (e.g: some cubes that weren't announced at the moment of writing are replicated...). If it is unsolvable, it would fail. On the contrary, it would retry the operation one more time.
- In the **optimizing** part, the keeper is called properly. 

And also a new function is added to the Metadata Manager:
- `hasConflicts` checks if there has been any conflict on the commit log.

### CODE REVIEW

**Comment by osopardo1** [comment]
I am closing this for the moment. It will be reopened when working on productization.

**Comment by osopardo1** [comment]
I reopen again this PR since we need to integrate with the Keeper/Curator properly.

File: src/main/scala/io/qbeast/spark/table/IndexedTable.scala [hunk 1]
```diff
@@ -193,16 +195,30 @@ private[table] class IndexedTableImpl(
 
   private def write(data: DataFrame, indexStatus: IndexStatus, append: Boolean): BaseRelation = {
     val revision = indexStatus.revision
-
-    if (exists) {
-      keeper.withWrite(tableID, revision.revisionID) { write =>
-        val announcedSet = write.announcedCubes.map(revision.createCubeId)
+    keeper.withWrite(tableID, revision.revisionID) { write =>
+      var tries = 2
```
Comment (polsm91) [inline src/main/scala/io/qbeast/spark/table/IndexedTable.scala:199]: Should it be a config property? It may make sense to tune it on different scenarios

**Review by polsm91 — APPROVED** [review]
In general looks good to me. Regarding testing, could we add a unit test which conflicts and another that does not, and call the `isConflicted` method?

File: src/main/scala/io/qbeast/spark/table/IndexedTable.scala (hunk 1 above)
Comment (osopardo1) [reply src/main/scala/io/qbeast/spark/table/IndexedTable.scala:199]: Mm, makes sense. Perhaps something like `spark.qbeast.keeper.numOfTries`?

File: core/src/main/scala/io/qbeast/core/model/MetadataManager.scala [hunk 2]
```diff
@@ -47,4 +47,20 @@ trait MetadataManager[DataSchema, FileDescriptor] {
...
+   * This function checks if there's a conflict. A conflict happens if there
+   * are new cubes that have been optimized but they were not announced.
+   *
+   * @param tableID the table ID
+   * @param revisionID the revision ID
+   * @param knownAnnounced the cubes we know they were announced when the write operation started.
+   * @param oldReplicatedSet the old replicated set
+   * @return true if there a conflict, false otherwise
+   */
+  def isConflicted(
```
[the remark this answers is not in the document]
Comment (osopardo1) [reply core/src/main/scala/io/qbeast/core/model/MetadataManager.scala:60]: `hasConflicts` better :+1:

**Pull request merged**